File: examples/typescript/other/custom_vfs.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { createHash } from "node:crypto";
import { eexist } from "@struktoai/mirage-core/utils/errors";
import { mkdtempSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import {
  Accessor,
  command,
  CommandSpec,
  ContentType,
  eisdir,
  enoent,
  enotdir,
  FileStat,
  FileType,
  GenericVFS,
  IOResult,
  MountMode,
  type PathSpec,
  registerVfsFactory,
  VFSAdapter,
  Workspace,
} from "@struktoai/mirage-node";

// A whole custom backend in one script: core functions over your
// data source, a read adapter with optional writes, one GenericVFS. Every generic
// command (ls, cat, grep, find, head, wc, ...) works for free, and so
// does versioning, in the shape the content calls for: the wiki's pages
// are the VFS's own, so they ride its state and a snapshot rebuilds
// the mount through the registered name with the pages as they were; the
// feed's live in a service, so the VFS is only observed and a load
// asks for it back rather than restoring a copy.

const ENC = new TextEncoder();
const DEC = new TextDecoder();

type Tree = { [name: string]: Tree | string };

const PAGES: Tree = {
  guides: {
    "quickstart.md": "# Quickstart\nMount anything as a filesystem.\n",
    "deploy.md": "# Deploy\nShip the gateway behind HTTP.\n",
  },
  "notes.md": "Remember: agents just speak bash.\n",
};

class WikiAccessor extends Accessor {
  constructor(
    public pages: Tree,
    readonly knownSizes = true,
  ) {
    super();
  }
}

function node(pages: Tree, key: string): Tree | string {
  let current: Tree | string = pages;
  for (const part of key.split("/").filter((p) => p !== "")) {
    // The errno constructors stamp `code`, which is what the command
    // chokepoints read to answer in GNU's voice: a bare `new Error('ENOENT')`
    // would surface as an internal failure instead.
    if (typeof current === "string") throw enoent(key);
    const child: Tree | string | undefined = current[part];
    if (child === undefined) throw enoent(key);
    current = child;
  }
  return current;
}

function readdir(accessor: WikiAccessor, path: PathSpec): Promise<string[]> {
  const found = node(accessor.pages, path.vfsPath);
  if (typeof found === "string") throw enotdir(path);
  const parent = path.virtual.replace(/\/+$/, "");
  return Promise.resolve(
    Object.entries(found).map(
      ([name, child]) =>
        `${parent}/${name}${typeof child === "string" ? "" : "/"}`,
    ),
  );
}

function readBytes(
  accessor: WikiAccessor,
  path: PathSpec,
): Promise<Uint8Array> {
  const found = node(accessor.pages, path.vfsPath);
  if (typeof found !== "string") throw eisdir(path);
  return Promise.resolve(ENC.encode(found));
}

function stat(accessor: WikiAccessor, path: PathSpec): Promise<FileStat> {
  const found = node(accessor.pages, path.vfsPath);
  const trimmed = path.virtual.replace(/\/+$/, "");
  const name = trimmed.slice(trimmed.lastIndexOf("/") + 1) || "/";
  if (typeof found !== "string")
    return Promise.resolve(
      new FileStat({ name, size: null, type: FileType.DIRECTORY }),
    );
  const data = ENC.encode(found);
  // The fingerprint is the content's own hash: the stable identity a
  // snapshot records for every read and a load checks for drift.
  const fingerprint = createHash("sha256")
    .update(data)
    .digest("hex")
    .slice(0, 16);
  return Promise.resolve(
    new FileStat({
      name,
      size: accessor.knownSizes ? data.length : null,
      type: FileType.FILE,
      content: ContentType.TEXT,
      fingerprint,
    }),
  );
}

function write(
  accessor: WikiAccessor,
  path: PathSpec,
  data: Uint8Array,
): Promise<void> {
  const parts = path.vfsPath.split("/").filter((p) => p !== "");
  const name = parts.pop() ?? "";
  let current: Tree = accessor.pages;
  for (const part of parts) {
    const next = (current[part] ??= {});
    if (typeof next === "string") throw enotdir(path);
    current = next;
  }
  current[name] = DEC.decode(data);
  return Promise.resolve();
}

function mkdir(
  accessor: WikiAccessor,
  path: PathSpec,
  parents = false,
): Promise<void> {
  const parts = path.vfsPath.split("/").filter((p) => p !== "");
  let current: Tree = accessor.pages;
  for (const [i, part] of parts.entries()) {
    const leaf = i === parts.length - 1;
    let next = current[part];
    if (next === undefined) {
      if (!leaf && !parents) throw enoent(path);
      next = current[part] = {};
    } else if (leaf && (!parents || typeof next === "string")) {
      throw eexist(path);
    }
    if (typeof next === "string") throw enotdir(path);
    current = next;
  }
  if (parts.length === 0 && !parents) throw eexist(path);
  return Promise.resolve();
}

// Optional: a bespoke domain verb, registered alongside the generics.
const wikiTitles = command({
  name: "wiki_titles",
  vfs: "wiki",
  spec: new CommandSpec(),
  fn: (accessor) => {
    const pages = (accessor as WikiAccessor).pages;
    const titles = ["guides/quickstart.md", "guides/deploy.md"].flatMap(
      (page) =>
        String(node(pages, page))
          .split("\n")
          .filter((line) => line.startsWith("# "))
          .map((line) => line.slice(2)),
    );
    return [ENC.encode(`${titles.join("\n")}\n`), new IOResult()];
  },
});

function makeIO(writable = true): VFSAdapter<WikiAccessor> {
  return new VFSAdapter({
    read: { readdir, readBytes, stat },
    ...(writable ? { writes: { write, mkdir } } : {}),
  });
}

class WikiVFS extends GenericVFS<WikiAccessor> {
  readonly wiki: WikiAccessor;

  constructor(pages: Tree = PAGES) {
    // A copy, so a live workspace and one loaded from its snapshot never
    // share pages.
    const wiki = new WikiAccessor(structuredClone(pages));
    super({
      name: "wiki",
      accessor: wiki,
      io: makeIO(),
      prompt: "A team wiki rendered as markdown files.",
      commands: wikiTitles,
      supportsSnapshot: true,
    });
    this.wiki = wiki;
  }

  // The pages are this VFS's own content, not a remote backend's,
  // so they ride its state: a snapshot or a version rebuilds the mount
  // with them and nothing has to be handed back by hand. A backend over
  // a remote service keeps the default state instead and is only
  // observed, through the fingerprints its stat reports.
  override getState(): { type: string; pages: Tree } {
    return { type: this.kind, pages: structuredClone(this.wiki.pages) };
  }

  // Typed against the saved shape the loader hands back (`type` plus
  // whatever getState wrote), which is what makes the class a VFS.
  override loadState(state: { type: string; pages?: Tree }): void {
    this.wiki.pages = structuredClone(state.pages ?? {});
  }
}

// The other half of the design: content that lives in a remote service.
// This object stands in for that service. The VFS holds no copy of
// it and keeps the default state, which says the mount has to be handed
// back live, so a snapshot pins what it read (through the fingerprints
// stat reports) and a load without the live VFS is refused rather
// than answered with an empty mount. Nothing registers it: a mount that
// is handed back needs no name in any registry.
const FEED: Tree = { "status.md": "All systems go.\n" };

class FeedVFS extends GenericVFS<WikiAccessor> {
  constructor() {
    super({
      name: "feed",
      accessor: new WikiAccessor(FEED, false),
      io: makeIO(false),
      prompt: "A status feed rendered as markdown.",
      supportsSnapshot: true,
    });
  }
}

async function show(ws: Workspace, line: string, prompt = "$"): Promise<void> {
  const io = await ws.shell(line);
  const out =
    io.exitCode === 0
      ? io.stdoutText
      : `${io.stderrText}exit ${String(io.exitCode)}\n`;
  console.log(`${prompt} ${line}\n${out}`);
}

async function main(): Promise<void> {
  // Registered up front: the name is what a snapshot rebuilds the mount
  // through, the same way workspace config names it.
  registerVfsFactory("wiki", () => Promise.resolve(new WikiVFS()));
  const ws = new Workspace(
    {
      "/wiki/": new WikiVFS(),
      "/nested/wiki/": [new WikiVFS(), MountMode.READ],
      "/feed/": new FeedVFS(),
    },
    { mode: MountMode.WRITE },
  );

  for (const line of [
    "ls /wiki/guides",
    "cat /wiki/notes.md",
    "grep -r Quickstart /wiki/",
    "find /wiki -name '*.md'",
    "wc -l /wiki/guides/quickstart.md",
    "wiki_titles",
    "cat /wiki/missing.md",
    "cat /feed/status.md",
    "cat /wiki/guides/*.md",
    "cat /nested/wiki/notes.md",
    "echo changed > /nested/wiki/notes.md",
    "cat /nested/wiki/notes.md",
    "wc -c /feed/status.md",
    "rm /feed/status.md",
    "gzip -c /feed/status.md | gunzip",
    "mkdir /wiki/new/child",
    "test ! -e /wiki/new && echo no-partial-parent",
    "mkdir -p /wiki/new/child",
    "mkdir /wiki/new/child",
    "mkdir -p /wiki/new/child",
    "mkdir -p /wiki/notes.md/child",
    "mkdir /wiki/guides/empty",
    "cp -r /wiki/guides /wiki/copied",
    "ls /wiki/copied",
    "cat /wiki/copied/quickstart.md",
    "rm -r /wiki/copied /wiki/notes.md",
    "rm -d /wiki/copied/empty /wiki/notes.md",
    "cp -r /nested/wiki/guides /nested/wiki/copied",
  ]) {
    await show(ws, line);
  }

  // The pages are versioned state: write one, snapshot, change it, and
  // the workspace loaded from the snapshot serves the page as it was
  // while the live one keeps the change. The registered name rebuilds
  // /wiki/ with no override; /feed/ is only observed, so a load that does
  // not hand it back is refused, naming the mount, and never restores it
  // as an empty directory.
  await show(ws, "echo '# Runbook' > /wiki/runbook.md");
  const tar = join(mkdtempSync(join(tmpdir(), "wiki-")), "wiki.tar");
  await ws.snapshot(tar);
  await show(ws, "echo '# Runbook, revised' > /wiki/runbook.md");
  let refused = false;
  try {
    await Workspace.load(tar);
  } catch (err) {
    if (!(err instanceof Error) || !err.message.includes("/feed/")) throw err;
    refused = true;
    console.log("load: refused, /feed/ must be handed back live");
  }
  if (!refused) throw new Error("a load without /feed/ was not refused");
  const restored = await Workspace.load(tar, {}, { "/feed/": new FeedVFS() });
  console.log("load: ok with /feed/ handed back\n");
  await show(ws, "cat /wiki/runbook.md", "live$");
  await show(restored, "cat /wiki/runbook.md", "restored$");
  await show(restored, "cat /feed/status.md", "restored$");

  await restored.close();
  await ws.close();
}

main().catch((err: unknown) => {
  console.error(err);
  process.exit(1);
});

File: examples/python/other/custom_vfs.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import hashlib
import tempfile
from copy import deepcopy
from pathlib import Path

from mirage import (NULL_INDEX, Accessor, CommandSpec, ContentType, FileStat,
                    FileType, GenericVFS, IndexCacheStore, IOResult, MountMode,
                    PathSpec, ReadOps, VFSAdapter, Workspace, WriteOps,
                    command, register_vfs)

# A whole custom backend in one script: four async core functions over
# your data source, a read adapter with optional writes, one GenericVFS. Every
# generic command (ls, cat, grep, find, head, wc, ...) works for free,
# and so does versioning, in the shape the content calls for: the wiki's
# pages are the VFS's own, so they ride its state and a snapshot
# rebuilds the mount through the registered name with the pages as they
# were; the feed's live in a service, so the VFS is only observed
# and a load asks for it back rather than restoring a copy.

PAGES = {
    "guides": {
        "quickstart.md": "# Quickstart\nMount anything as a filesystem.\n",
        "deploy.md": "# Deploy\nShip the gateway behind HTTP.\n",
    },
    "notes.md": "Remember: agents just speak bash.\n",
}


class WikiAccessor(Accessor):

    def __init__(self, pages: dict, known_sizes: bool = True) -> None:
        self.pages = pages
        self.known_sizes = known_sizes


def _node(pages: dict, key: str):
    node = pages
    for part in [p for p in key.split("/") if p]:
        if not isinstance(node, dict) or part not in node:
            raise FileNotFoundError(key)
        node = node[part]
    return node


async def readdir(
    accessor: WikiAccessor,
    path: PathSpec,
    index: IndexCacheStore = NULL_INDEX,
) -> list[str]:
    node = _node(accessor.pages, path.vfs_path)
    if not isinstance(node, dict):
        raise NotADirectoryError(path.virtual)
    parent = path.virtual.rstrip("/")
    return [
        f"{parent}/{name}" + ("/" if isinstance(child, dict) else "")
        for name, child in node.items()
    ]


async def read_bytes(
    accessor: WikiAccessor,
    path: PathSpec,
    index: IndexCacheStore = NULL_INDEX,
) -> bytes:
    node = _node(accessor.pages, path.vfs_path)
    if isinstance(node, dict):
        raise IsADirectoryError(path.virtual)
    return node.encode()


async def stat(
    accessor: WikiAccessor,
    path: PathSpec,
    index: IndexCacheStore = NULL_INDEX,
) -> FileStat:
    node = _node(accessor.pages, path.vfs_path)
    name = path.virtual.rstrip("/").rsplit("/", 1)[-1] or "/"
    if isinstance(node, dict):
        return FileStat(name=name, size=None, type=FileType.DIRECTORY)
    data = node.encode()
    # The fingerprint is the content's own hash: the stable identity a
    # snapshot records for every read and a load checks for drift.
    return FileStat(name=name,
                    size=len(data) if accessor.known_sizes else None,
                    type=FileType.FILE,
                    content=ContentType.TEXT,
                    fingerprint=hashlib.sha256(data).hexdigest()[:16])


async def write(accessor: WikiAccessor, path: PathSpec, data: bytes) -> None:
    *folders, name = (p for p in path.vfs_path.split("/") if p)
    node = accessor.pages
    for part in folders:
        node = node.setdefault(part, {})
        if not isinstance(node, dict):
            raise NotADirectoryError(path.virtual)
    node[name] = data.decode()


async def mkdir(accessor: WikiAccessor,
                path: PathSpec,
                parents: bool = False) -> None:
    node = accessor.pages
    for part in (p for p in path.vfs_path.split("/") if p):
        node = node.setdefault(part, {})
        if not isinstance(node, dict):
            raise NotADirectoryError(path.virtual)


# Optional: a bespoke domain verb, registered alongside the generics.
@command("wiki_titles", vfs="wiki", spec=CommandSpec())
async def wiki_titles(accessor, paths, texts, opts):
    titles = [
        line[2:] for page in ("guides/quickstart.md", "guides/deploy.md")
        for line in _node(accessor.pages, page).splitlines()
        if line.startswith("# ")
    ]
    return ("\n".join(titles) + "\n").encode(), IOResult()


def make_io(*, writable: bool = True) -> VFSAdapter:
    return VFSAdapter(
        read=ReadOps(readdir=readdir, read_bytes=read_bytes, stat=stat),
        writes=WriteOps(write=write, mkdir=mkdir) if writable else WriteOps(),
    )


class WikiVFS(GenericVFS):
    """The backend as a class, so the registry can build it by name."""

    def __init__(self, pages: dict | None = None) -> None:
        # A copy, so a live workspace and one loaded from its snapshot
        # never share pages.
        self.wiki = WikiAccessor(deepcopy(PAGES if pages is None else pages))
        super().__init__(
            name="wiki",
            accessor=self.wiki,
            io=make_io(),
            prompt="A team wiki rendered as markdown files.",
            commands=[wiki_titles],
            supports_snapshot=True,
        )

    # The pages are this VFS's own content, not a remote backend's,
    # so they ride its state: a snapshot or a version rebuilds the mount
    # with them and nothing has to be handed back by hand. A backend over
    # a remote service keeps the default state instead and is only
    # observed, through the fingerprints its stat reports.
    def get_state(self) -> dict:
        return {"type": self.name, "pages": deepcopy(self.wiki.pages)}

    def load_state(self, state: dict) -> None:
        self.wiki.pages = deepcopy(state.get("pages", {}))


# The other half of the design: content that lives in a remote service.
# This dict stands in for that service. The VFS holds no copy of
# it and keeps the default state, which says the mount has to be handed
# back live, so a snapshot pins what it read (through the fingerprints
# stat reports) and a load without the live VFS is refused rather
# than answered with an empty mount. Nothing registers it: a mount that
# is handed back needs no name in any registry.
FEED = {"status.md": "All systems go.\n"}


class FeedVFS(GenericVFS):

    def __init__(self) -> None:
        super().__init__(name="feed",
                         accessor=WikiAccessor(FEED, known_sizes=False),
                         io=make_io(writable=False),
                         prompt="A status feed rendered as markdown.",
                         supports_snapshot=True)


async def show(ws: Workspace, line: str, prompt: str = "$") -> None:
    result = await ws.shell(line)
    out = await result.stdout_str()
    if result.exit_code != 0:
        out = f"{await result.stderr_str()}exit {result.exit_code}\n"
    print(f"{prompt} {line}\n{out}", flush=True)


async def main():
    # Registered up front: the name is what a snapshot rebuilds the
    # mount through, the same way workspace YAML names it.
    register_vfs("wiki", WikiVFS)
    ws = Workspace(
        {
            "/wiki/": WikiVFS(),
            "/nested/wiki/": (WikiVFS(), MountMode.READ),
            "/feed/": FeedVFS()
        },
        mode=MountMode.WRITE)

    for line in (
            "ls /wiki/guides",
            "cat /wiki/notes.md",
            "grep -r Quickstart /wiki/",
            "find /wiki -name '*.md'",
            "wc -l /wiki/guides/quickstart.md",
            "wiki_titles",
            "cat /wiki/missing.md",
            "cat /feed/status.md",
            "cat /wiki/guides/*.md",
            "cat /nested/wiki/notes.md",
            "echo changed > /nested/wiki/notes.md",
            "cat /nested/wiki/notes.md",
            "wc -c /feed/status.md",
            "rm /feed/status.md",
            "gzip -c /feed/status.md | gunzip",
            "mkdir /wiki/guides/empty",
            "cp -r /wiki/guides /wiki/copied",
            "ls /wiki/copied",
            "cat /wiki/copied/quickstart.md",
            "rm -r /wiki/copied /wiki/notes.md",
            "rm -d /wiki/copied/empty /wiki/notes.md",
            "cp -r /nested/wiki/guides /nested/wiki/copied",
    ):
        await show(ws, line)

    # The pages are versioned state: write one, snapshot, change it, and
    # the workspace loaded from the snapshot serves the page as it was
    # while the live one keeps the change. The registered name rebuilds
    # /wiki/ with no override; /feed/ is only observed, so a load that
    # does not hand it back is refused, naming the mount, and never
    # restores it as an empty directory.
    await show(ws, "echo '# Runbook' > /wiki/runbook.md")
    tar = Path(tempfile.mkdtemp(prefix="wiki-")) / "wiki.tar"
    await ws.snapshot(tar)
    await show(ws, "echo '# Runbook, revised' > /wiki/runbook.md")
    try:
        await Workspace.load(tar)
    except ValueError as exc:
        assert "/feed/" in str(exc)
        print("load: refused, /feed/ must be handed back live", flush=True)
    else:
        raise AssertionError("a load without /feed/ was not refused")
    restored = await Workspace.load(tar, mounts={"/feed/": FeedVFS()})
    print("load: ok with /feed/ handed back\n", flush=True)
    await show(ws, "cat /wiki/runbook.md", prompt="live$")
    await show(restored, "cat /wiki/runbook.md", prompt="restored$")
    await show(restored, "cat /feed/status.md", prompt="restored$")

    await restored.close()
    await ws.close()


asyncio.run(main())
